cloud/shared/bin/lib/terraform.py

```diff
@@ -1,37 +1,118 @@
 import subprocess
 import os
+import sys
+import re
 import shutil
 import shlex
+import inspect
 from typing import Optional
 
 from cloud.shared.bin.lib.config_loader import ConfigLoader
 from cloud.shared.bin.lib.print import print
+from cloud.aws.templates.aws_oidc.bin.aws_cli import AwsCli
 
 
-# TODO(#2741): When using this for Azure make sure to setup backend bucket prior to calling these functions.
-def perform_apply(
+def force_unlock(
         config_loader: ConfigLoader,
-        is_destroy=False,
-        terraform_template_dir: Optional[str] = None):
-    '''Generates terraform variable files and runs terraform init and apply.'''
+        lock_id: str,
+        terraform_template_dir: Optional[str] = None,
+        initialize=True):
     if not terraform_template_dir:
         terraform_template_dir = config_loader.get_template_dir()
-    tf_vars_filename = config_loader.tfvars_filename
 
-    terraform_cmd = f'terraform -chdir={terraform_template_dir}'
+    if initialize:
+        perform_init(
+            config_loader, terraform_template_dir, False)  # upgrade = False
+
+    terraform_cmd = f'terraform -chdir={terraform_template_dir} force-unlock -force {lock_id}'
+    print(f" - Run {terraform_cmd}")
+    subprocess.check_call(shlex.split(terraform_cmd))
+
+
+def perform_init(
+        config_loader: ConfigLoader,
+        terraform_template_dir: Optional[str] = None,
+        upgrade: bool = True):
+    if not terraform_template_dir:
+        terraform_template_dir = config_loader.get_template_dir()
+
+    init_cmd = f'terraform -chdir={terraform_template_dir} init'
+    if upgrade:
+        init_cmd += ' -upgrade'
 
     if config_loader.use_local_backend:
-        print(' - Run terraform init -upgrade -reconfigure')
-        subprocess.check_call(
-            shlex.split(f'{terraform_cmd} init -upgrade -reconfigure'))
+        init_cmd += ' -reconfigure'
     else:
-        print(' - Run terraform init -upgrade')
-        init_cmd = f'{terraform_cmd} init -input=false -upgrade'
+        init_cmd += ' -input=false'
         # backend vars file can be absent when pre-terraform setup is running
         if os.path.exists(os.path.join(terraform_template_dir,
                                        config_loader.backend_vars_filename)):
             init_cmd += f' -backend-config={config_loader.backend_vars_filename}'
-        subprocess.check_call(shlex.split(init_cmd))
+    print(f" - Run {init_cmd}")
+    output, exit_code = capture_stderr(init_cmd)
+    if exit_code > 0:
+        is_tty = sys.stdin.isatty()
```

Comment:
**nitpick:** Maybe add a comment above saying this is checking for interactive mode?